### ISSUE TITLE
Poor performance when updating thousands of data

### ISSUE THREAD

**Issue by risalfajar** [grahamearley/FirestoreGoogleAppsScript#168]
## Explain the Problem in Detail
I have thousands of items that I want to update.
Currently, there's only `updateDocument()` API available, which only updates a single data. Since Apps Script doesn't support parallelization (CMIIW), our only option is to call `updateDocument()` multiple times **synchronously**, which is very bad for performance.

Do you have any solution/workaround for this?

##### Library Version: 33

**Comment by abkarino** [comment]
You would need to use a `batchWrite` operation which would encapsulate up to 500 writes.
However, this feature is not available in this repo yet. There is a Pull Request already open for that.

**Comment by Michael-Ackerman3956** [comment]
when is the batchWrite function coming along? I am also having the same problem. I need to run hundreds of the app script function...

**Comment by abkarino** [comment]
I have no clue.
https://github.com/grahamearley/FirestoreGoogleAppsScript/pull/144
I was just checking if there has been any updates recently in this repo.

**Comment by Michael-Ackerman3956** [comment]
Just posted my issue in SO - https://stackoverflow.com/questions/77289105/create-or-update-firestore-collection-based-on-google-app-script

**Comment by abkarino** [comment]
I already have that feature deployed on my fork, you can use it from there or you can do a new fork and deploy it to your own script.

**Comment by risalfajar** [comment]
I don't think this library is still maintained, the last release is three years ago.

@grahamearley would you confirm on this?

**Comment by MichaelJoo** [comment]
> I already have that feature deployed on my fork, you can use it from there or you can do a new fork and deploy it to your own script.

I am now trying to use your fork, but still not sure how I can use your version. If I just your code, I would get an error saying "TypeError: firestore.batch is not a function
"

**Comment by Michael-Ackerman3956** [comment]
> I already have that feature deployed on my fork, you can use it from there or you can do a new fork and deploy it to your own script.

Installed your V4 library and it works like a charm :)

**Comment by LaughDonor** [comment]
> I don't think this library is still maintained, the last release is three years ago.
> 
> @grahamearley would you confirm on this?

I apologize on the lack of maintenance. I have scheduled time to work on this within the coming week to address and catch up to all the issues. I'm aware of @abkarino's changes and should be incorporating them into a release as well. Thanks for your patience! :)

**Comment by risalfajar** [comment]
@LaughDonor It has been almost 1 year, do you have any problem?

**Comment by Michael-Ackerman3956** [comment]
Any updates on this? I am now importing @abkarino 's batch function again, which works well.

**Comment by ywc8891** [comment]
> Any updates on this? I am now importing [@abkarino](https://github.com/abkarino) 's batch function again, which works well.

The script cannot be accessed since last week. [@abkarino](https://github.com/abkarino)

**Comment by MichaelJoo** [comment]
@ywc8891 I was able to run it again, you just have to remove the previous library and add it again. Just to be safe, I also downloaded the forked version at local. We are all still waiting for @grahamearley to merge this update. I believe once the batch updates are merged, the library is far more robust.

**Comment by Michael-Ackerman3956** [comment]
@grahamearley @LaughDonor The batch function library is not accessible, although we were able to download the files to local. Would there be any update to merge this major "batch update" functions into the existing library?? 

@abkarino Any issues at your repo? I downloaded it to local, but the library doesn't work anymore. Looking forward for resolution :))

Anyway, I worked around it by forking @abkarino 's repo and creating my own local library. But I hope to use the main version in future as my implementations are rather crude and I am not very good at JS/TS.

**Comment by abkarino** [comment]
I am terribly sorry for the broken library on my end. 
I made a mistake back then and the code was deployed under the company Google domain and I couldn't migrate it to personal. I didn't want to change the id for existing users but I left the company and they might have limited the access to that library.
I can communicate with them or reupload this time to my personal account tomorrow. 
I developed it for the company use but released to the public hoping it would have been to the repo. 
I also wanted to add support for transactions but I didn't do that as the repo hasn't been maintained for over a year or more.

**Comment by Michael-Ackerman3956** [comment]
@abkarino To be honest, I have been using your repo to a great use and it is a huge time savor, esp when my forte is not in java/google app script. I was able to workaround by forking your repo and creating my own app script library. 

But I would love to use your personal account version, if it would be maintained and improved further. Also, thank you for being a good owner and maintaining it.

**Comment by abkarino** [comment]
@MichaelJoo I appreciate your kind words, thank you!
I redeployed the script into Gmail domain. I also got the original library back up for public use for existing users and hopefully they will migrate into the Gmail one. I would have fixed it earlier if I was notified.
